### PR TITLE
 Fetch access code

### PR DESCRIPTION
**Pull request by chrisrcoles**
## Description

Final PR that will activate access codes. See [design doc](https://docs.google.com/document/d/1D5f52XMC0PMdBaiv85I5W9nMxOTv0EU1mFsBMDqSXo0/edit) for more detailed context. 

## Reviewer Notes

This reopens [PR #2287](https://github.com/transcom/mymove/pull/2287).

## Setup

Make sure .envrc.local has REQUIRES_ACCESS_CODE set to true.

```
1. make server_run
2. make client_run
3. go run cmd/generate_access_codes/main.go -ppm 20 -hhg 20
4. login as new Milmove user 
```

## Code Review Verification Steps

* [x] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/tree/master/docs/backend.md#logging)
* [x] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/tree/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:
  * [x] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](https://github.com/transcom/mymove/tree/master/docs/database.md#zero-downtime-migrations))
  * [x] Have been communicated to #dp3-engineering - _Message to be sent 7/8 AM when deploy happens_
* [x] There are no aXe warnings for UI.
* [x] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
* [x] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [x] User facing changes have been reviewed by design.
* [x] Request review from a member of a different team.
* [x] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/166649949)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -883,8 +883,8 @@ workflows:
             - acceptance_tests_local
           # if testing on experimental, you can disable these tests by using the commented block below.
           # filters:
-          #  branches:
-          #    ignore: placeholder_branch_name
+          #   branches:
+          #     ignore: placeholder_branch_name
```
Comment (chrisgilmerproj) [inline .circleci/config.yml]: This change should be undone. If the spacing is out of alignment for the next person CircleCI might fail on them.

File: .circleci/config.yml [hunk 2]
```diff
@@ -676,7 +676,7 @@ jobs:
       - APP_ENVIRONMENT: 'experimental'
     steps:
       - deploy_app_client_tls_steps:
-          compare_host: "" # leave blank since we want experimental to be able to roll back
+          compare_host: '' # leave blank since we want experimental to be able to roll back
```
Comment (chrisgilmerproj) [inline .circleci/config.yml]: What happened here? Is there a yaml formatter that changed this?

File: cmd/milmove/serve.go [hunk 3]
```diff
@@ -129,8 +129,11 @@ func initServeFlags(flag *pflag.FlagSet) {
 	// Verbose
 	cli.InitVerboseFlags(flag)
 
-	// Sort command line flags
-	flag.SortFlags = true
+	// Feature Flags
+	cli.InitFeatureFlag(flag)
+
+	// Don't sort flags
+	flag.SortFlags = false
```
Comment (chrisgilmerproj) [inline cmd/milmove/serve.go]: Is there a reason why this is turned off?  I'd prefer to keep it on and have good naming on the flags.

File: pkg/cli/featureflag.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package cli
+
+import (
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+)
+
+const (
+	//RequiresAccessCode flag
+	RequiresAccessCode string = "requires-access-code"
```
Comment (chrisgilmerproj) [inline pkg/cli/featureflag.go]: My preference here would be to use the name space `feature-flag`. So this:
 
```suggestion
	FeatureFlagAccessCode string = "feature-flag-access-code"
```

File: pkg/cli/featureflag.go [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+)
+
+// InitFeatureFlag initializes FeatureFlags command line flags
+func InitFeatureFlag(flag *pflag.FlagSet) {
+	flag.Bool(RequiresAccessCode, false, "Flag (bool) to enable requires-access-code")
+}
+
+// CheckFeatureFlag validates Verbose command line flags
+func CheckFeatureFlag(v *viper.Viper) error {
```
Comment (chrisgilmerproj) [inline pkg/cli/featureflag.go]: Please still add a test file for this file even if there's nothing being checked.  The file would look like: https://github.com/transcom/mymove/blob/master/pkg/cli/dps_test.go

File: config/app.container-definition.json [hunk 6]
```diff
@@ -187,6 +187,10 @@
     {
       "name": "SERVE_SWAGGER_UI",
       "value": "{{ .SERVE_SWAGGER_UI }}"
+    },
+    {
+      "name": "REQUIRES_ACCESS_CODE",
+      "value": "{{ .REQUIRES_ACCESS_CODE }}"
```
Comment (chrisgilmerproj) [inline config/app.container-definition.json]:
```suggestion
      "value": "{{ .FEATURE_FLAG_ACCESS_CODE }}"
```

File: migrations/20190422163230_create_access_code.up.sql [hunk 7]
```diff
@@ -1,8 +1,12 @@
...
+	service_member_id uuid UNIQUE REFERENCES service_members(id),
 	code text UNIQUE NOT NULL,
 	move_type text NOT NULL,
-	created_at timestamp with time zone NOT NULL DEFAULT NOW(),
-	claimed_at timestamp with time zone
+	created_at timestamp
+	with time zone NOT NULL DEFAULT NOW
+	(),
+	claimed_at timestamp
+	with time zone
```
Comment (chrisgilmerproj) [inline migrations/20190422163230_create_access_code.up.sql]: So you're in a bit of a mess here if you've already run this on Experimental. Because modifying this file won't modify the experimental DB and will get us out of alignment. To add `UNIQUE` as an index you'll need to add a new migration file with a new name. Then we'll be certain that Experimental is the same as the other environments.

File: migrations/20190510205156_add_requires_access_code_to_service_members.up.fizz [hunk 8]
```diff
@@ -0,0 +1 @@
+add_column("service_members", "requires_access_code", "bool", {"default": false})
```
Comment (chrisgilmerproj) [inline migrations/20190510205156_add_requires_access_code_to_service_members.up.fizz]: Setting `{"default": false}` is going to lock the service member table while its added. I'd prefer that we added the column with no default (ie allowing null) and then come back and update each row individually so we don't lock.  Once that's done we can alter the table to set `NOT NULL` on it.

File: migrations/20190422163230_create_access_code.up.sql (hunk 7 above)
Comment (chrisgilmerproj) [reply migrations/20190422163230_create_access_code.up.sql]: Actually I'd prefer this file not be modified at all, even with the new formatting. I'm pretty sure it's already been deployed to staging and production.

Comment (kahlouie) [reply migrations/20190422163230_create_access_code.up.sql]: I agree with @chrisgilmerproj 's reason on the second comment. I'd suggest if the `claimed_at` needs to have time zone, we should create a new migration to update it.

Comment (chrisrcoles) [reply migrations/20190422163230_create_access_code.up.sql]: I agree with both of you - the only change here we are doing is adding the unique attribute to `service_member_id` so no other migration is needed if we decide to drop this change as @chrisgilmerproj suggests. I'll remove it and update the line formatting.

**Comment by chrisrcoles** [comment]
@kahlouie - thanks for the review and the suggestion. So I definitely have some feelings on what you brought up. This code has been designed, planned, tested on experimental, has been code reviewed numerous times, and tested extensively. Things will undoubtedly go wrong in engineering but when they do, we should respond swiftly and appropriately. While having the benefit of type 2 decision making, creating that PR implies that we do not have confidence in the code that we are releasing are operating in an environment of fear - both of which are not healthy. If you have some reason why you are not confident in the release of that code, let's deal with that first.

File: migrations/20190422163230_create_access_code.up.sql (hunk 7 above)
Comment (chrisrcoles) [reply migrations/20190422163230_create_access_code.up.sql]: `claimed_at` already has the timezone - it's a line formatting difference that you're seeing

File: migrations/20190510205156_add_requires_access_code_to_service_members.up.fizz (hunk 8 above)
Comment (chrisrcoles) [reply migrations/20190510205156_add_requires_access_code_to_service_members.up.fizz]: Ok to clarify your point, I need to:
1. remove `require_access_code`’s `{"default": false}` from the migration
2. create another PR that just adds updates each row to set default to false and not null or is it okay to do this other migration in the same PR?

File: cmd/milmove/serve.go [hunk 9]
```diff
@@ -129,7 +129,10 @@ func initServeFlags(flag *pflag.FlagSet) {
 	// Verbose
 	cli.InitVerboseFlags(flag)
 
-	// Sort command line flags
+	// Feature Flags
+	cli.InitFeatureFlag(flag)
+
+	// Don't sort flags
```
Comment (chrisgilmerproj) [inline cmd/milmove/serve.go]: You'll want to revert the comment too :)

File: pkg/cli/featureflag_test.go [hunk 10]
```diff
@@ -0,0 +1,6 @@
+package cli
+
+func (suite *cliTestSuite) TestCheckFeatureFlag() {
+	suite.Setup(InitDPSFlags, []string{})
```
Comment (chrisgilmerproj) [inline pkg/cli/featureflag_test.go]:
```suggestion
	suite.Setup(InitFeatureFlagsFlag, []string{})
```

File: pkg/cli/featureflag.go [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+	"github.com/spf13/viper"
+)
+
+const (
+	//RequiresAccessCode flag
+	FeatureFlagAccessCode string = "feature-flag-access-code"
+)
+
+// InitFeatureFlag initializes FeatureFlags command line flags
+func InitFeatureFlag(flag *pflag.FlagSet) {
```
Comment (chrisgilmerproj) [inline pkg/cli/featureflag.go]: Usually this is pluralized:

```suggestion
func InitFeatureFlags(flag *pflag.FlagSet) {
```

File: config/env/experimental.env [hunk 12]
```diff
@@ -13,3 +13,4 @@ DB_SSL_MODE=verify-full
 DB_SSL_ROOT_CERT=/bin/rds-combined-ca-bundle.pem
 DEVLOCAL_CA=/config/tls/devlocal-ca.pem
 DEVLOCAL_AUTH=false
+REQUIRES_ACCESS_CODE=false
```
Comment (chrisgilmerproj) [inline config/env/experimental.env]:
```suggestion
FEATURE_FLAG_ACCESS_CODE=false
```

File: config/env/prod.env [hunk 13]
```diff
@@ -13,3 +13,4 @@ DB_SSL_MODE=verify-full
 DB_SSL_ROOT_CERT=/bin/rds-combined-ca-bundle.pem
 DEVLOCAL_CA=/config/tls/devlocal-ca.pem
 DEVLOCAL_AUTH=false
+REQUIRES_ACCESS_CODE=true
```
Comment (chrisgilmerproj) [inline config/env/prod.env]:
```suggestion
FEATURE_FLAG_ACCESS_CODE=true
```

File: config/env/staging.env [hunk 14]
```diff
@@ -13,3 +13,4 @@ DB_SSL_MODE=verify-full
 DB_SSL_ROOT_CERT=/bin/rds-combined-ca-bundle.pem
 DEVLOCAL_CA=/config/tls/devlocal-ca.pem
 DEVLOCAL_AUTH=false
+REQUIRES_ACCESS_CODE=false
```
Comment (chrisgilmerproj) [inline config/env/staging.env]:
```suggestion
FEATURE_FLAG_ACCESS_CODE=false
```

File: migrations/20190422163230_create_access_code.up.sql [hunk 15]
```diff
@@ -1,4 +1,5 @@
-CREATE TABLE access_codes (
+CREATE TABLE access_codes
+(
```
Comment (chrisgilmerproj) [inline migrations/20190422163230_create_access_code.up.sql]: I'd still revert this change

File: pkg/handlers/contexts.go [hunk 16]
```diff
@@ -47,6 +47,8 @@ type HandlerContext interface {
 	SetSendProductionInvoice(sendProductionInvoice bool)
 	UseSecureCookie() bool
 	SetUseSecureCookie(useSecureCookie bool)
+	SetFeatureFlags(flags ...FeatureFlag)
```
Comment (chrisgilmerproj) [inline pkg/handlers/contexts.go]: Why not just `SetFeatureFlag`? It seems like you run the risk of having some piece of code totally rewrite all the flags if all they want to do is set one flag. Alternatively, why not have `GetFeatureFlags`? Allowing someone to get all of them? This API feels like it ought to either do pluralized or singular actions but not one of each.

Comment (chrisrcoles) [reply pkg/handlers/contexts.go]: agreed! I've updated this!

File: cmd/milmove/serve.go [hunk 17]
```diff
@@ -129,7 +129,10 @@ func initServeFlags(flag *pflag.FlagSet) {
 	// Verbose
 	cli.InitVerboseFlags(flag)
 
-	// Sort command line flags
+	// Feature Flags
+	cli.InitFeatureFlags(flag)
+
+	// Sort flags
```
Comment (chrisgilmerproj) [inline cmd/milmove/serve.go]: I think the original comment here was better.
```suggestion
	// Sort command line flags
```
